pkg/sql/set_cluster_setting_test.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package sql_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+var testSetting = settings.RegisterIntSetting(
+	"sql.test_setting",
+	"a test setting used by TestSetClusterSetting",
+	0,
+)
+
+// TestSetClusterSetting verifies the expected semantics for the SET CLUSTER
+// SETTING statement. The update should be visible on the node that executes the
+// statement as soon as the statement returns, and it should eventually become
+// visible on the other nodes in the cluster too.
+func TestSetClusterSetting(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	const numNodes = 2
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, numNodes, base.TestClusterArgs{})
+	defer tc.Stopper().Stop(ctx)
+
+	sv0 := &tc.Server(0).ClusterSettings().SV
+	sv1 := &tc.Server(0).ClusterSettings().SV
+	verify := func(e int64) {
+		t.Helper()
+		if a := testSetting.Get(sv0); e != a {
+			t.Fatalf("settings update was not immediately reflected on node that executed update: "+
+				"expected %d, got %d", e, a)
+		}
+		testutils.SucceedsSoon(t, func() error {
+			if a := testSetting.Get(sv1); e != a {
+				return fmt.Errorf("settings update is not yet reflected on other node in cluster: "+
+					"expected %d, got %d", e, a)
+			}
+			return nil
+		})
+	}
+
+	db0 := tc.ServerConn(0)
+	if _, err := db0.Exec(`SET CLUSTER SETTING sql.test_setting = 1`); err != nil {
```

Comment:
Should this put the `SET` inside an explicit txn and verify on sv1 that it does _not_ appear until the commit? 

This will still make for a potentially surprising UX when your `SET` doesn't show up inside your own txn though prior to commit. Ugh, maybe these should not be allowed in txns.

Reply:
Yeah, agreed, these probably shouldn't be allowed in txns.

Actually, are you comfortable with this change going in without a test? This test sends up being flaky because multiple updates in quick succession will race with the gossip update and the test gets confused. But it really does completely eliminate the flakiness in every test that runs `SET CLUSTER SETTING kv.range_merge.queue_enabled = false`. Really trying to get this change in in time for next week's beta. 😬

Reply:
Hm. Been pondering this one for the last hour and I'm not crazy about it, particularly now there are two ways the in-mem value is managed instead of just being a cache of what's been gossiped. I'm trying to figure out what i'd like better though... playing with some ideas locally -- thinking some way that the `SET` could wait up to some timeout hoping to see its write come back via gossip and be processed. Thinking something like a channel pub/sub thing or updates handled, revision counter on each setting, or even just polling the setting value.

Reply:
played with this a bit yesterday and came up with a slight variation -- wait for the new value to be returned when accessing the setting. What do you think about #29063?